pmlc/rt/level_zero/level_zero_utils.cc

```diff
@@ -0,0 +1,502 @@
+  context_ = context;
+  if (event_pool_ == nullptr) {
+    ze_event_pool_desc_t descriptor = {};
+    descriptor.stype = ZE_STRUCTURE_TYPE_EVENT_POOL_DESC;
+
+    descriptor.pNext = nullptr;
+    descriptor.flags = flags;
+    descriptor.count = count;
+
+    EXPECT_EQ(ZE_RESULT_SUCCESS, zeEventPoolCreate(context, &descriptor, 0,
+                                                   nullptr, &event_pool_));
+    EXPECT_NE(nullptr, event_pool_);
+
+    pool_indexes_available_.resize(count, true);
+  }
+}
+
+void zeEventPool::create_event(ze_event_handle_t &event,
+                               ze_event_scope_flags_t signal,
+                               ze_event_scope_flags_t wait) {
+  // Make sure the event pool is initialized to at least defaults:
+  InitEventPool(context_, 32);
+  ze_event_desc_t desc = {};
+  memset(&desc, 0, sizeof(desc));
+  desc.stype = ZE_STRUCTURE_TYPE_EVENT_DESC;
+  desc.pNext = nullptr;
+  desc.signal = signal;
+  desc.wait = wait;
+  event = nullptr;
+  desc.index = -1;
+  for (uint32_t i = 0; i < pool_indexes_available_.size(); i++) {
+    if (pool_indexes_available_[i]) {
+      desc.index = i;
+      break;
+    }
+  }
+  EXPECT_EQ(ZE_RESULT_SUCCESS, zeEventCreate(event_pool_, &desc, &event));
+  EXPECT_NE(nullptr, event);
+  handle_to_index_map_[event] = desc.index;
+  pool_indexes_available_[desc.index] = false;
+}
+
+void zeEventPool::destroy_event(ze_event_handle_t event) {
+  std::map<ze_event_handle_t, uint32_t>::iterator it =
+      handle_to_index_map_.find(event);
+
+  EXPECT_NE(it, handle_to_index_map_.end());
+  pool_indexes_available_[(*it).second] = true;
+  handle_to_index_map_.erase(it);
+  EXPECT_EQ(ZE_RESULT_SUCCESS, zeEventDestroy(event));
+}
+
+void append_barrier(ze_command_list_handle_t cl, ze_event_handle_t hSignalEvent,
+                    uint32_t numWaitEvents, ze_event_handle_t *phWaitEvents) {
+  EXPECT_EQ(ZE_RESULT_SUCCESS,
+            zeCommandListAppendBarrier(cl, hSignalEvent, numWaitEvents,
+                                       phWaitEvents));
+}
+
+// Group
+void set_group_size(ze_kernel_handle_t hFunction, uint32_t groupSizeX,
+                    uint32_t groupSizeY, uint32_t groupSizeZ) {
+  EXPECT_EQ(ZE_RESULT_SUCCESS, zeKernelSetGroupSize(hFunction, groupSizeX,
+                                                    groupSizeY, groupSizeZ));
+}
+
+void suggest_group_size(ze_kernel_handle_t hFunction, uint32_t globalSizeX,
+                        uint32_t globalSizeY, uint32_t globalSizeZ,
+                        uint32_t &groupSizeX, uint32_t &groupSizeY,
+                        uint32_t &groupSizeZ) {
+  EXPECT_EQ(ZE_RESULT_SUCCESS,
+            zeKernelSuggestGroupSize(hFunction, globalSizeX, globalSizeY,
+                                     globalSizeZ, &groupSizeX, &groupSizeY,
+                                     &groupSizeZ));
+}
+
+// Helper
+std::string to_string(const ze_result_t result) {
+  if (result == ZE_RESULT_SUCCESS) {
+    return "ZE_RESULT_SUCCESS";
```

Comment:
A macro could really shorten this function. Like:

```
#define ERRSTR(_err_) if (result == _err_) return #_err_;

ERRSTR(ZE_RESULT_SUCCESS);
ERRSTR(ZE_RESULT_NOT_READY);
...
```


```

Reply:
Done.